clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/AmazonInstanceTypeCachingAgent.java

```diff
@@ -0,0 +1,196 @@
+import java.io.InputStream;
+import java.net.URI;
+import java.util.*;
+import java.util.stream.Collectors;
+
+public class AmazonInstanceTypeCachingAgent implements CachingAgent {
+
+  private static final TypeReference<Map<String, Object>> ATTRIBUTES = new TypeReference<Map<String, Object>>() {
+  };
+
+  // https://pricing.us-east-1.amazonaws.com/offers/v1.0/aws/AmazonEC2/current/us-west-2/index.json
+  private final String region;
+  private final AccountCredentialsRepository accountCredentialsRepository;
+  private final URI pricingUri;
+  private final HttpHost pricingHost;
+  private final HttpClient httpClient;
+  private final ObjectMapper objectMapper = new ObjectMapper().disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
+
+
+  public AmazonInstanceTypeCachingAgent(String region, AccountCredentialsRepository accountCredentialsRepository) {
+    this(region, accountCredentialsRepository, HttpClients.createDefault());
+  }
+
+  //VisibleForTesting
+  AmazonInstanceTypeCachingAgent(String region, AccountCredentialsRepository accountCredentialsRepository, HttpClient httpClient) {
+    this.region = region;
+    this.accountCredentialsRepository = accountCredentialsRepository;
+    pricingHost = HttpHost.create("https://pricing.us-east-1.amazonaws.com");
+    pricingUri = URI.create("/offers/v1.0/aws/AmazonEC2/current/" + region + "/index.json");
+    this.httpClient = httpClient;
+  }
+
+  @Override
+  public Collection<AgentDataType> getProvidedDataTypes() {
+    return Collections.unmodifiableList(Arrays.asList(new AgentDataType(Keys.Namespace.INSTANCE_TYPES.getNs(), AgentDataType.Authority.AUTHORITATIVE), new AgentDataType(getAgentType(), AgentDataType.Authority.AUTHORITATIVE)));
+  }
+
+  @Override
+  public CacheResult loadData(ProviderCache providerCache) {
+    try {
+      Set<String> matchingAccounts = accountCredentialsRepository.getAll()
+        .stream()
+        .filter(AmazonCredentials.class::isInstance)
+        .map(AmazonCredentials.class::cast)
+        .filter(ac -> ac.getRegions().stream().anyMatch(r -> region.equals(r.getName())))
+        .map(AccountCredentials::getName)
+        .collect(Collectors.toSet());
+
+      if (matchingAccounts.isEmpty()) {
+        return new DefaultCacheResult(Collections.emptyMap());
+      }
+
+      CacheData metadata = providerCache.get(getAgentType(), "metadata", RelationshipCacheFilter.none());
+      MetadataAttributes metadataAttributes = null;
+      if (metadata != null) {
+        metadataAttributes = objectMapper.convertValue(metadata.getAttributes(), MetadataAttributes.class);
+      }
+
+      Set<String> instanceTypes = null;
+      if (metadataAttributes != null && metadataAttributes.etag != null && metadataAttributes.cachedInstanceTypes != null) {
+
+        //we have enough from a previous request to not re-request if the etag is unchanged..
+        HttpResponse headResponse = httpClient.execute(pricingHost, new HttpHead(pricingUri));
```

Comment:
out of curiosity, what are the default connect timeouts for stock `HttpClient` ... anything to be concerned with there?